GcExtensions.md

```diff
@@ -0,0 +1,353 @@
+# Garbage Collector Extensions
+
+- **Title:** Garbage collector extensions for better foreign language support
+- **Author:** Reimer Behrends <behrends@gmail.com>
+- **Created:** May 2018
+- **Status:** work in progress
+
+## Introduction
+
+The support for modules written entirely or partly in foreign languages
+to interface with the Julia GC or to use the GC for allocations that do
+not neatly fit Julia's type system or use low-level approaches not
+available in Julia (such as irregular data structure layouts) is
+currently somewhat limited.
+
+This proposal aims at fleshing out the API for allowing more complex
+interaction of foreign code with the GC, especially the use of long-lived
+foreign objects that are inextricably interwoven with Julia
+objects.
+
+Specific use cases that we are trying to address are:
+
+* *Allowing the Julia GC to manage foreign objects with arbitrary
+  layouts.* Not all objects -- especially those from preexisting
+  libraries -- fit Julia's type system, for example, specialized
+  container types written in C/C++. Such objects can comprise multiple
+  memory blocks that require a custom marking mechanism and may also
+  require finalizers written in C.
+* *Providing additional roots to the GC.* Currently, to have additional
+  roots, they must be stored in a location that is visible to Julia.
+  This can be expensive if such roots are updated frequently or are
+  contained in data structures that would have to be laboriously
+  translated into a format usable by Julia. Instead, we want to allow
+  for roots to be discoverable at the beginning of a garbage collection.
+* *Conservative scanning of stack frames and objects.* Currently,
+  scanning does have to be precise. If we desire to use the GC for
+  foreign code that requires conservative scanning (especially for
+  foreign stack frames), then it is necessary to have functionality
+  that determines whether a machine word is a pointer to an object,
+  including to its interior.
```

Comment:
It'll be better to have the three functions separated.

Custom layout and additional roots are useful features in general.
Conservative scanning is also fine, as long as it's clear it's up to the user to make sure their compiler doesn't mess things up and that it's up to the user to implement all the layouts supported by the julia GC.

Reply:
They are already mostly separate, I'd say. At least I am not aware of anything in the API that would preclude you from only using one, or two, of them. If you see issues, we'd appreciate if you could give us a hint as to where. Thanks!

Reply:
It's not about this julep. I just mean that adding each of these separately is better, both for the commit (history) and for PR.

Reply:
Ahhh, OK. So you are saying that if/when we submit our code as a PR, we should have this in three separate commits. I guess we could do that, gotta think about it (but again, we'll wait a bit for the dust to settle ;-).

Reply:
Note that this split has now been done, I hope it's now the way you wanted it.